src/ray/stats/metric_exporter_client_test.cc

```diff
@@ -0,0 +1,108 @@
+// limitations under the License.
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+#include <chrono>
+#include <iostream>
+#include <thread>
+#include <vector>
+
+#include "absl/memory/memory.h"
+#include "ray/stats/metric_exporter.h"
+#include "ray/stats/metric_exporter_client.h"
+#include "ray/stats/stats.h"
+
+namespace ray {
+using namespace stats;
+
+class MockExporterClient1 : public MetricExporterDecorator {
+ public:
+  MockExporterClient1(std::shared_ptr<MetricExporterClient> exporter)
+      : MetricExporterDecorator(exporter) {
+    client1_count = 0;
+  }
+  void ReportMetrics(const MetricPoints &points) override {
+    MetricExporterDecorator::ReportMetrics(points);
+    client1_count += points.size();
+    client1_value = points.back().value;
+    RAY_LOG(INFO) << "Client 1 " << client1_count << " " << points.back().value;
+  }
+  static int GetCount() { return client1_count; }
+  static int GetValue() { return client1_value; }
+
+ private:
+  static int client1_count;
+  static int client1_value;
+};
+
+class MockExporterClient2 : public MetricExporterDecorator {
+ public:
+  MockExporterClient2(std::shared_ptr<MetricExporterClient> exporter)
+      : MetricExporterDecorator(exporter) {
+    client2_count = 0;
+  }
+  void ReportMetrics(const MetricPoints &points) override {
+    MetricExporterDecorator::ReportMetrics(points);
+    client2_count += 2 * points.size();
+    RAY_LOG(INFO) << "Client 2 " << client2_count << " " << points.back().value;
+    client2_value = points.back().value;
+  }
+  static int GetCount() { return client2_count; }
+  static int GetValue() { return client2_value; }
+
+ private:
+  static int client2_count;
+  static int client2_value;
+};
+
+class MetricExporterClientTest : public ::testing::Test {
+ public:
+  void SetUp() {
+    const stats::TagsType global_tags = {{stats::LanguageKey, "CPP"},
+                                         {stats::WorkerPidKey, "1000"}};
+    ray::stats::Init("127.0.0.1:8888", global_tags, false);
+    std::shared_ptr<MetricExporterClient> exporter(new stats::StdoutExporterClient());
+    std::shared_ptr<MetricExporterClient> mock1(new MockExporterClient1(exporter));
+    std::shared_ptr<MetricExporterClient> mock2(new MockExporterClient2(mock1));
+    MetricExporter::Register(mock2);
+  }
+
+  void Shutdown() {}
+};
+
+int MockExporterClient1::client1_count;
+int MockExporterClient2::client2_count;
+int MockExporterClient1::client1_value;
+int MockExporterClient2::client2_value;
+
+TEST_F(MetricExporterClientTest, decorator_test) {
+  for (size_t i = 0; i < 100; ++i) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+    stats::CurrentWorker().Record(i + 1);
+  }
+  std::this_thread::sleep_for(std::chrono::milliseconds(10000));
+  ASSERT_GE(100, MockExporterClient1::GetValue());
+  ASSERT_GE(100, MockExporterClient2::GetValue());
+  ASSERT_EQ(1, MockExporterClient1::GetCount());
+  ASSERT_EQ(2, MockExporterClient2::GetCount());
+}
```

Comment:
Can we also test 
1. if number of points are bigger than the batch size, if the reportMetrics is triggered?
2. ExportToPoints(opencensus::stats::ViewData::DataMap<opencensus::stats::Distribution>)

Reply:
> Can we also test
> 
> 1. if number of points are bigger than the batch size, if the reportMetrics is triggered?
> 2. ExportToPoints(opencensus::stats::ViewData::DataMapopencensus::stats::Distribution)

All mentioned above tests have been added in suite.

Reply:
For this test, `std::this_thread::sleep_for(std::chrono::milliseconds(kReportFlushInterval + 200));` is unnecessary because metrics should've been emitted before 10 seconds are passed right? Can we remove it from this test?

Reply:
It's necessary to make sure opencensus will trigger a timer handler.

Reply:
> For this test, `std::this_thread::sleep_for(std::chrono::milliseconds(kReportFlushInterval + 200));` is unnecessary because metrics should've been emitted before 10 seconds are passed right? Can we remove it from this test?

It's necessary to make sure opencensus will trigger a timer handler.

Reply:
I think since we already tested the timer handler in the previous test, we just need to test if exporting is happening when the number of metrics exceed the batch size. In this case, we don't need to trigger the timer handler I believe? Did I miss something here?

Reply:
> Since we already tested the timer handler in the previous test, we just need to test if exporting is happening when the number of metrics exceed the batch size. In this case, we don't need to trigger the timer handler I believe? Did I miss something here?

In opencensus exporting mechanism, our Exporter Function code pathes will be executed if and only if timer handlers in registered list  are triggered periodically.

Reply:
Ah, I see. So, `ReportMetrics` function only records the metrics at the buffer? (I thought this function actually triggers export).

Reply:
> Ah, I see. So, `ReportMetrics` function only records the metrics at the buffer? (I thought this function actually triggers export).

In user thread : 1. User recording -> 2. View data from measurement
Timer handler thread : 1. Periodically collecting view data to register exporter.